### ISSUE TITLE
KeePassXC does not show in Alt+Tab under GNOME Wayland until StartupNotify times out

### ISSUE THREAD

**Issue by chenxiaolong** [keepassxreboot/keepassxc#6423]
## Overview

KeePass currently does not show in Alt+Tab under GNOME Wayland until `StartupNotify` (from the `.desktop` file) times out. I suspect this is because of the naming differences between the X11 `WM_CLASS` and Wayland/xdg-shell's `app_id`.

Under X11, I believe the Qt XCB platform plugin uses `QCoreApplication::applicationName()` to set the `WM_CLASS` property. KeePassXC's `WM_CLASS` is:
```
$ xprop WM_CLASS
WM_CLASS(STRING) = "keepassxc", "KeePassXC"
```

Under Wayland, qtwayland uses the `.desktop` filename to set the `app_id`: https://github.com/qt/qtwayland/blob/20d7cc45933a4a26c886b47c0b69da82195e7204/src/client/qwaylandwindow.cpp#L159-L163 Thus, the value is set to `org.keepassxc.KeePassXC`. This can be confirmed by using gnome-shell's "looking glass" feature: 
![image](https://user-images.githubusercontent.com/646253/115159923-a5df1880-a063-11eb-8794-66474867166a.png)

Because KeePassXC's `.desktop` file specifies:
```
StartupWMClass=keepassxc
```
it will never match under a Wayland session.

## Steps to Reproduce

1. Load Wayland gnome-shell session
2. Start KeePassXC via its `.desktop` file
3. Immediately try to Alt+Tab

## Expected Behavior

KeePassXC should immediately show up in the Alt+Tab list.

## Actual Behavior

KeePassXC is missing from Alt+Tab list until `StartupNotify` times out.

## Context

```
KeePassXC - Version 2.6.4
Revision: 34a78f0

Qt 5.15.2
Debugging mode is disabled.

Operating system: Fedora 34 (Workstation Edition)
CPU architecture: x86_64
Kernel: linux 5.10.19-200.fc33.x86_64

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- libgcrypt 1.9.2-unknown
```

Operating System: Fedora 34 (also reproducible under 33)
Desktop Env: Gnome
Windowing System: Wayland

**Comment by chenxiaolong** [comment]
I think this can be fixed by changing: https://github.com/keepassxreboot/keepassxc/blob/805574cac1c5295b37541c67d150eaf247e845b2/src/main.cpp#L62

to call:
```cpp
    Application::setApplicationName("org.keepassxc.KeePassXC");
```
and the changing the `.desktop` file to:
```
StartupWMClass=org.keepassxc.KeePassXC
```
Another option would be to rename the `.desktop` file to `keepassxc.desktop`.

Both solutions should work under both X11 and Wayland, but I'm not sure if either would have undesired side effects.

**Comment by droidmonkey** [comment]
This is stupid, but thank you for finding this problem.

**Comment by droidmonkey** [comment]
I think just changing StartupWMClass to the reverse DNS pattern should be sufficient. Does X11 even care about that attribute?

**Comment by chenxiaolong** [comment]
> I think just changing StartupWMClass to the reverse DNS pattern should be sufficient. Does X11 even care about that attribute?

Good question. I'll try that out really quick.

**Comment by chenxiaolong** [comment]
By only making the:
```
StartupWMClass=org.keepassxc.KeePassXC
```
change, it works properly in Wayland. KeePassXC shows up in Alt-Tab immediately. On X11, KeePassXC still immediately shows up in Alt+Tab, but gnome-shell's looking glass shows the window as untracked:

![image](https://user-images.githubusercontent.com/646253/115161416-8815b180-a06b-11eb-9b7e-04ba79abee59.png)

likely because `WM_CLASS` (`"KeePassXC"`) now no longer matches `StartupWMClass`. I suspect this breaks `StartupNotify` behavior on X11.

**Comment by droidmonkey** [comment]
We can also just do `QGuiApplication::setDesktopFileName("keepassxc")` which will use the common WMClass name on Wayland instead of Qt building it dynamically.

**Comment by chenxiaolong** [comment]
I can test that out really quickly. Do you know if that would have other side effects in Qt?

**Comment by droidmonkey** [comment]
The docs state it is only used in this specific use case:
```
This property holds the base name of the desktop entry for this application

This is the file name, without the full path, of the desktop entry that represents this application according to the freedesktop desktop entry specification.

This property gives a precise indication of what desktop entry represents the application and it is needed by the windowing system to retrieve such information without resorting to imprecise heuristics.

The latest version of the freedesktop desktop entry specification can be obtained here.

This property was introduced in Qt 5.7.
```

**Comment by droidmonkey** [comment]
Might there also be a better attribute to put in the desktop file for Wayland's use? I feel like this would be a problem for a lot of applications.

**Comment by chenxiaolong** [comment]
I wasn't able to find a better attribute unfortunately, but maybe I missed something. This would indeed be a problem for a lot of applications, but it seems most Qt applications don't meet all these conditions:

* `StartupNotify` exists
* `StartupWMClass` == X11's `WM_CLASS`
* `.desktop` filename (and this Wayland's `app_id`) != X11's `WM_CLASS`

The only other program I have installed that behaves this way is KDE's Konsole.

**Comment by chenxiaolong** [comment]
I looked through gnome-shell's code a bit and I think this is a GNOME bug. If the Wayland `app_id` (or X11 `WM_CLASS`) doesn't match `StartupWMClass`, it's supposed to try and match the name with the `.desktop` file: https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/40.0/src/shell-window-tracker.c#L192-212

I still don't know if `StartupWMClass` is meant to be used under Wayland or if this naming mismatch affects other desktop environments.

**Comment by chenxiaolong** [comment]
Yep, looks like this is a known issue on GNOME + Wayland: https://gitlab.gnome.org/GNOME/mutter/-/issues/1330

I think this can be closed, unless KeePassXC would be willing to work around this issue `StartupNotify=false`.

**Comment by chenxiaolong** [comment]
Looks like startup notifications will be properly supported by a new `xdg_activation` Wayland protocol that will be implemented by toolkits and desktop environments.

I'm going to close this issue since KeePassXC is not doing anything incorrectly.

**Comment by droidmonkey** [comment]
Great! Now I'm upset I deleted my comment blaming Gnome itself.

**Comment by Gert-dev** [comment]
FWIW, there has been an update on this from GNOME's side, see also [this ticket](https://gitlab.gnome.org/GNOME/mutter/-/issues/2689), especially [this reply](https://gitlab.gnome.org/GNOME/mutter/-/issues/2689#note_1696224) and [this reply](https://gitlab.gnome.org/GNOME/mutter/-/issues/1330#note_1696148).

If I understood correctly, this means that since the previous posts, `xdg_activation` is now supported by GNOME with `StartupNotify=true` on Wayland, but the issue won't be fixed until KeePassXC can upgrade to Qt 6.2, which is related to #7774.
